# SAE S2.02 — Exploration algorithmique d'un problème

## Partie 1 : PageRank — version itérative, premier exemple

---

**1. Justifier pourquoi l'application de l'algorithme de la puissance itérée à la matrice `Q` permet de calculer le score de chacune des pages.** 

L'algorithme PageRank simule la navigation d’un Homme sur le web créé sous forme de matrice avec les vecteurs représentant les liens hypertextes du web. À chaque itération, les pages transmettent une partie de leur score aux pages vers lesquelles elles pointent. Les pages recevant de nombreux liens voient leur score augmenter de façon non proportionnelle car en fonction du score de la page qui pointe vers la page recevante, elle reçoit plus ou moins de score en fonction du score de la précédente donc, plus elle est importante, plus la page pointée le devient également. <br>

Après plusieurs itérations, les scores se stabilisent et forment un vecteur r qui est le score des vecteurs et, est solution de l’équation r=Qr. Ce vecteur représente alors l'importance relative de chaque page du graphe. <br>



**2. Implémenter cet algorithme pour calculer le score de chacune des pages du graphe précédent. On vérifiera (numériquement) que le vecteur de score obtenu est bien approximativement solution de `r = Qr`.**

In [14]:
import numpy as np

def somme_col(M):
    #fais la somme de chaque colonne
    return np.sum(M, axis=0)

def stochastique(M): # prend en parametre une matrice d'adjacence transposée
    new = np.zeros((M.shape[0], M.shape[0]), dtype=float)
    s = somme_col(M)
    for i in range(len(M)):
        for j in range(len(M[i])):
            if s[j] != 0:
                new[i][j] = M[i][j] / s[j] # divise chaque element d'une colonne par la somme de celle ci
            else:
                new[i][j] = 0
    return new

def norme(X):
    return np.sqrt(np.sum(X**2))


def puissance_iterée(M, e):
    old = np.random.randint(0, 10, M.shape[0]) # crée un vecteur aléatoire
    tmp = old.dot(M)  # multiplie le vecteur a la matrice
    new = tmp / norme(tmp) # AXn / ||AXn||
    i = 0 # compteur pour indiquer le nombre d'iteration
    while norme(new - old) > e:
        i += 1
        if i == 10000:   # seuil d'arrêt en cas de boucle infinie
            return new, norme(M.dot(new))
        old = new
        tmp = M.dot(old)
        new = tmp / norme(tmp)
    return new, norme(M.dot(new)), i





Partie1 = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
], dtype=float)

e = 10**(-10)


Q = stochastique(Partie1.T)

r, val, i= puissance_iterée(Q, e)

# Affiche le score des pages de la matrice web.
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")

print("\nr =")
print(r)
print("\nQr =")
print(Q.dot(r))
print("\nNombre d'iteration = ", i)
print("\nValeur propre = ", val)
print("\nMarge d'erreur =", norme(r - Q.dot(r)))
print("\nVu que la marge d'erreur est tres petite on peut dire que r est approximativement solution de r = Qr")

Page 1  : 0.4377
Page 2  : 0.0875
Page 3  : 0.1167
Page 4  : 0.1459
Page 5  : 0.1897
Page 6  : 0.5252
Page 7  : 0.1751
Page 8  : 0.3502
Page 9  : 0.1751
Page 10 : 0.4377
Page 11 : 0.0875
Page 12 : 0.1167
Page 13 : 0.1459
Page 14 : 0.1897

r =
[0.43768811 0.08753762 0.11671683 0.14589604 0.18966485 0.52522573
 0.17507524 0.35015049 0.17507524 0.43768811 0.08753762 0.11671683
 0.14589604 0.18966485]

Qr =
[0.43768811 0.08753762 0.11671683 0.14589604 0.18966485 0.52522573
 0.17507524 0.35015049 0.17507524 0.43768811 0.08753762 0.11671683
 0.14589604 0.18966485]

Nombre d'iteration =  218

Valeur propre =  1.0000000000000002

Marge d'erreur = 8.585322179748825e-11

Vu que la marge d'erreur est tres petite on peut dire que r est approximativement solution de r = Qr


**3. Analyser qualitativement la pertinence du résultat obtenu.** <br>

Avec les resultats precedant que l'on peut voir que la page avec le plus de socre est la page 6, c'est a dire la page qui a été le plus reference par des pages elles même très référencé.

## Partie 2 : PageRank — version itérative, deuxième exemple

---

**1.Appliquer l'algorithme de la Partie 1 au graphe suivant et commenter le résultat obtenu.**

In [20]:
Partie2 = np.array([
    [0,0,0,0,0],
    [1,0,0,0,0],
    [1,0,0,1,0],
    [1,1,0,0,0],
    [1,1,0,0,0]
], dtype=float)


Q = stochastique(Partie2.T)

r, val, i = puissance_iterée(Q, e)


/tmp/ipykernel_146882/1763890319.py:32: RuntimeWarning: invalid value encountered in divide
  new = tmp / norme(tmp)


L'algorithme crash car il y a un puit dans le graphe ce qui fait qu'il y a une colonne de 0 une fois transposé dans la matrice faisant une division par 0.

**2. En utilisant cette matrice de transition (avec `α = 0,85`), calculer les scores de chacune des pages du graphe précédent. On vérifiera (numériquement) que le vecteur de score obtenu est bien approximativement solution de `r = Pr`. Analyser qualitativement la pertinence du résultat obtenu.**


In [21]:
def transition(M, a) :
    N = len(M)
    P = np.zeros((N, N), dtype=float)
    s = somme_col(M)
    for i in range(N) :
        for j in range(N) :
            if s[j] == 0 :
                P[i][j] = 1/N
            else :
                P[i][j] = a * M[i][j] + (1- a) / N
    return P

web2 = np.array([
    [0,0,0,0,0],
    [1,0,0,0,0],
    [1,0,0,1,0],
    [1,1,0,0,0],
    [1,1,0,0,0]
], dtype=float)

a = 0.85

Q = stochastique(web2.T)

P = transition(Q, a)

r, val, i = puissance_iterée(P, e)


print("r =")
print(r)
print("\nPr =")
print(P.dot(r))
print("\nValeur propre = ", val)
print("\nMarge d'erreur =", norme(r - P.dot(r)))
print("\nVu que la marge d'erreur est tres petite on peut dire que r est approximativement solution de r = Pr")

r =
[0.82582244 0.40102189 0.19748693 0.28141887 0.19748693]

Pr =
[0.82582244 0.40102189 0.19748693 0.28141887 0.19748693]

Valeur propre =  1.0000000000066442

Marge d'erreur = 2.2144880164054182e-11

Vu que la marge d'erreur est tres petite on peut dire que r est approximativement solution de r = Pr


## Partie 3 : PageRank — version itérative, analyse des paramètres

Pour le moment on pose α = 0,85 et on considère le graphe de la Partie 1.

**1.Analyser l'influence du critère d'arrêt dans l'algorithme de la puissance itérée.**

Le critère d'arrêt influence la durée, le nombre d'itérations de l'algorithme et la precision du résultat, plus il est petit plus l'algorithme va faire d'itération et le résultat seras précis.




In [22]:
web = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
], dtype=float)

a = 0.85
Q = stochastique(web.T)
P = transition(Q, a)

print("Influence du critère d'arrêt")
for e in [1e-2, 1e-4, 1e-6, 1e-8, 1e-10]:
    r, val, i = puissance_iterée(P, e)

    print("\n----------------------------------------------------------------------")
    print("\nPrécision =", e)
    print("\nErreur :", norme(r - P.dot(r)))
    print("\nNombre d'iteration :", i)
    print("\n")
    for page in range(len(r)):
        print(f"Page {page+1:<3}: {r[page]:.4f}")


Influence du critère d'arrêt

----------------------------------------------------------------------

Précision = 0.01

Erreur : 0.00704384448677774

Nombre d'iteration : 9


Page 1  : 0.4802
Page 2  : 0.1179
Page 3  : 0.1516
Page 4  : 0.1831
Page 5  : 0.2304
Page 6  : 0.4438
Page 7  : 0.1608
Page 8  : 0.2979
Page 9  : 0.1608
Page 10 : 0.4483
Page 11 : 0.1108
Page 12 : 0.1420
Page 13 : 0.1709
Page 14 : 0.2141

----------------------------------------------------------------------

Précision = 0.0001

Erreur : 7.187080757684482e-05

Nombre d'iteration : 23


Page 1  : 0.4623
Page 2  : 0.1141
Page 3  : 0.1464
Page 4  : 0.1763
Page 5  : 0.2214
Page 6  : 0.4481
Page 7  : 0.1625
Page 8  : 0.3006
Page 9  : 0.1625
Page 10 : 0.4619
Page 11 : 0.1140
Page 12 : 0.1463
Page 13 : 0.1762
Page 14 : 0.2212

----------------------------------------------------------------------

Précision = 1e-06

Erreur : 6.502107489131094e-07

Nombre d'iteration : 35


Page 1  : 0.4621
Page 2  : 0.1141
Page 3  : 0.14

**2. Ajouter quelques hubs (pages qui ont beaucoup de liens sortants) et autorités (pages qui ont beaucoup de liens entrants). Commenter l'impact sur les scores**

In [23]:
# 3.2A - Ajout des Hubs seulement
webAvecHub = np.array([
    [0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0],  # Page 1 [Hub] -> {2, 3, 4, 5, 6}
    [1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1],  # Page 10 [Hub] -> {6, 9, 11, 12, 13, 14}
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
])

e = 10**(-10)
a = 0.85
Q = stochastique(webAvecHub.T)
P = transition(Q, a)

print("Influence du critère d'arrêt sur Hub seulement")
r, val, i = puissance_iterée(P, e)

print("\nPrécision =", e)
print("\nMarge d'erreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
print("\n")
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")

print("\n--------------------------------------------------")

# 3.2B - Ajout d'Autorités seulement
webAvecAutorite = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],  # Page 6 [Autorité] -> {8}
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  # Page 14 [Autorité] -> {}
])

e = 10**(-10)
a = 0.85
Q = stochastique(webAvecAutorite.T)
P = transition(Q, a)

print("Influence du critère d'arrêt sur Authorite seulement")
r, val, i= puissance_iterée(P, e)

print("\nPrécision =", e)
print("\nMarge d'erreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
print("\n")
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")

print("\n---------------------------------------------------")

# 3.2C - Ajout des Hubs et des Autorités
webAvecHubAutorite = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 1 [Autorité] -> {2, 6}
    [1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1],  # Page 6 [Hub] -> {1, 7, 8, 9, 10}
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 8 [Autorité] -> {}
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1],
    [1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0]  # Page 14 [Hub] -> {10, 11, 13}
])


e = 10**(-10)
a = 0.85
Q = stochastique(webAvecHubAutorite.T)
P = transition(Q, a)

print("Influence du critère d'arrêt sur Hub et Authorite")
r, val, i= puissance_iterée(P, e)

print("\nPrécision =", e)
print("\nMarge d'erreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
print("\n")
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")


Influence du critère d'arrêt sur Hub seulement

Précision = 1e-10

Marge d'erreur : 5.869380596346884e-11

Nombre d'iteration : 76


Page 1  : 0.4364
Page 2  : 0.0969
Page 3  : 0.1244
Page 4  : 0.1498
Page 5  : 0.1881
Page 6  : 0.4665
Page 7  : 0.2291
Page 8  : 0.3621
Page 9  : 0.2291
Page 10 : 0.4364
Page 11 : 0.0969
Page 12 : 0.1244
Page 13 : 0.1498
Page 14 : 0.1881

--------------------------------------------------
Influence du critère d'arrêt sur Authorite seulement

Précision = 1e-10

Marge d'erreur : 8.173171246919216e-11

Nombre d'iteration : 141


Page 1  : 0.3465
Page 2  : 0.0944
Page 3  : 0.1212
Page 4  : 0.1459
Page 5  : 0.1832
Page 6  : 0.6260
Page 7  : 0.0355
Page 8  : 0.5978
Page 9  : 0.0355
Page 10 : 0.1382
Page 11 : 0.0590
Page 12 : 0.0757
Page 13 : 0.0912
Page 14 : 0.1145

---------------------------------------------------
Influence du critère d'arrêt sur Hub et Authorite

Précision = 1e-10

Marge d'erreur : 5.0854526593737674e-11

Nombre d'iteration : 41


Page 1  :

**3. Essayez d'accroître le score de certaines pages. Expliquez votre méthode et validez\-la expérimentalement.**

Si on fait en sorte que toutes les pages pointes vers une page en particulier en ajoutant des liens entrants supplémentaires dans la matrice d'adjacence, on peut accroitre son score, dans ce cas-là, on accroit ba le score de la `page 1`.

In [24]:
webPageXScoreBoost = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 2 -> Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 3 -> Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 4 -> Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 5 -> Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],  # Page 6 -> Page 1, 10
    [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 7 -> Page 1
    [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 8 -> Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],  # Page 9 -> Page 10, 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Page 10 > Page 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],  # Page 11 -> Page 10, 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],  # Page 12 -> Page 10, 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],  # Page 13 -> Page 10, 1
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]   # Page 14 -> Page 10, 1
])

e = 10**(-10)
a = 0.85
Q = stochastique(webPageXScoreBoost.T)
P = transition(Q, a)
r, val, i= puissance_iterée(P, e)

print("\nPrécision =", e)
print("\nMarge d'rreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
print("\n")
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")


Précision = 1e-10

Marge d'rreur : 8.196787540608834e-11

Nombre d'iteration : 81


Page 1  : 0.9007
Page 2  : 0.1757
Page 3  : 0.1757
Page 4  : 0.1757
Page 5  : 0.1757
Page 6  : 0.1950
Page 7  : 0.0226
Page 8  : 0.0226
Page 9  : 0.0226
Page 10 : 0.1535
Page 11 : 0.0226
Page 12 : 0.0226
Page 13 : 0.0226
Page 14 : 0.0226


**4. Faites varier le facteur d'amortissement `α` pour analyser son influence. On rappelle que `α ∈ [0, 1]`.**

In [25]:

web = np.array([
    [0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
], dtype=float)

e = 10**(-10)
#a = 0.30
Q = stochastique(web.T)
#P = transition(Q, a)



for a in [0.1, 0.3, 0.6, 0.99]:
    P = transition(Q, a)
    r, val, i= puissance_iterée(P, e)
    print("\n----------------------------------------------------------------------")
    print("Calcul avec alpha = ", a)
    print("\nNombre d'itération :", i)
    print("\n")
    for page in range(len(r)):
        print(f"Page {page+1:<3}: {r[page]:.4f}")

print("\nOn peut voir que plus α est élevée plus les pages trés pointé monopolise le score.")


----------------------------------------------------------------------
Calcul avec alpha =  0.1

Nombre d'itération : 10


Page 1  : 0.3128
Page 2  : 0.2460
Page 3  : 0.2542
Page 4  : 0.2587
Page 5  : 0.2672
Page 6  : 0.2797
Page 7  : 0.2491
Page 8  : 0.2740
Page 9  : 0.2491
Page 10 : 0.3128
Page 11 : 0.2460
Page 12 : 0.2542
Page 13 : 0.2587
Page 14 : 0.2672

----------------------------------------------------------------------
Calcul avec alpha =  0.3

Nombre d'itération : 17


Page 1  : 0.3838
Page 2  : 0.2057
Page 3  : 0.2262
Page 4  : 0.2396
Page 5  : 0.2622
Page 6  : 0.3121
Page 7  : 0.2138
Page 8  : 0.2780
Page 9  : 0.2138
Page 10 : 0.3838
Page 11 : 0.2057
Page 12 : 0.2262
Page 13 : 0.2396
Page 14 : 0.2622

----------------------------------------------------------------------
Calcul avec alpha =  0.6

Nombre d'itération : 35


Page 1  : 0.4473
Page 2  : 0.1530
Page 3  : 0.1836
Page 4  : 0.2081
Page 5  : 0.2460
Page 6  : 0.3738
Page 7  : 0.1741
Page 8  : 0.2786
Page 9  : 0.1741

On observe que plus α est élevée plus les pages très pointé monopolise le score.

## Partie 4 : PageRank — version itérative, analyse des résultats

Dans cette partie, vous appliquerez l'algorithme de PageRank version itérative à différents graphes. L'objectif principal n'est pas uniquement de calculer les scores, mais d'**analyser et d'interpréter les classements obtenus** à la lumière de la structure du graphe étudié. On s'attachera notamment à expliquer pourquoi certains sommets obtiennent un score élevé ou faible, en s'appuyant sur leurs liens entrants, leurs liens sortants et leur position dans le réseau.

**1. En utilisant le logiciel d'exploration de site web présent sur la page AMeTICE (et vu en TP de R2.07), construire trois matrices de votre choix et appliquer l'algorithme de PageRank à ces matrices. L'ordre de ces matrices sera compris entre 10 et 30. Appliquer l'algorithme de PageRank à chacune d'elles et analyser les classements obtenus en les confrontant à la structure du site exploré. Les matrices, le résultat du logiciel d'exploration ainsi que le site web choisi devront apparaître clairement dans le dossier.**

### Premier site : Mickaël-Martin Nevot
```string
;extern;www.mickael\-martin\-nevot.com;www.mickael\-martin\-nevot.com/admin;www.mickael\-martin\-nevot.com/faq;www.mickael\-martin\-nevot.com/liens;www.mickael\-martin\-nevot.com/credits;www.mickael\-martin\-nevot.com/mentions\-legales;www.mickael\-martin\-nevot.com/plan\-du\-site;www.mickael\-martin\-nevot.com?c=home;www.mickael\-martin\-nevot.com/recueil\-des\-cours/sxx\-presentation\-du\-recueil\-des\-cours\-et\-de\-l\-architecture\-des\-documents.pdf;www.mickael\-martin\-nevot.com?c=dark;www.mickael\-martin\-nevot.com/faq?c=home;www.mickael\-martin\-nevot.com/faq?c=dark;www.mickael\-martin\-nevot.com/liens?c=home
extern;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/admin;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/faq;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/liens;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/credits;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/mentions\-legales;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com/plan\-du\-site;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com?c=home;0;1;0;0;0;0;0;0;0;0;0;0;0;0
www.mickael\-martin\-nevot.com/recueil\-des\-cours/sxx\-presentation\-du\-recueil\-des\-cours\-et\-de\-l\-architecture\-des\-documents.pdf;0;1;0;1;1;1;1;1;1;0;1;1;1;1
www.mickael\-martin\-nevot.com?c=dark;0;1;0;0;0;0;0;0;0;0;0;0;0;0
www.mickael\-martin\-nevot.com/faq?c=home;0;0;0;1;0;0;0;0;0;0;0;1;1;0
www.mickael\-martin\-nevot.com/faq?c=dark;0;0;0;1;0;0;0;0;0;0;0;1;1;0
[www.mickael\-martin\-nevot.com/liens?c=home;0;0;0;0;1;0;0;0;0;0;0;0;0;1](http://www.mickael-martin-nevot.com/liens?c=home;0;0;0;0;1;0;0;0;0;0;0;0;0;1)
```

Lien = https://www.mickael-martin-nevot.com/

In [31]:
nevot = np.array([
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,0,1,1,1,1,1,1,0,1,1,1,1],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,1,0,0,0,0,0,0,0,1,1,0],
    [0,0,0,1,0,0,0,0,0,0,0,1,1,0],
    [0,0,0,0,1,0,0,0,0,0,0,0,0,1]
])

e = 10**(-10)
a = 0.85
Q = stochastique(nevot.T)
P = transition(Q, a)
r, val, i= puissance_iterée(P, e)

print("\nMarge d'rreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
for page in range(len(r)):
    print(f"Page {page+1:<3}: {r[page]:.4f}")


Marge d'rreur : 4.4414589300322725e-11

Nombre d'iteration : 30
Page 1  : 0.0341
Page 2  : 0.4601
Page 3  : 0.0341
Page 4  : 0.3933
Page 5  : 0.2964
Page 6  : 0.1704
Page 7  : 0.1704
Page 8  : 0.1704
Page 9  : 0.1704
Page 10 : 0.0341
Page 11 : 0.1704
Page 12 : 0.3933
Page 13 : 0.3933
Page 14 : 0.2964


On peut voir que la page avec le meilleur score est la 2ème car c'est la page d'accueil \(on peut y revenir depuis toutes les autres pages du site\) et les autres sont hautes par exemple la 4 et la 5 qui sont FAQ et lien et qui sont référencé par la page d'accueil et la 12, et 13 qui sont le thème sombre de la FAQ et le lien vers home de la page lien qui ont tous deux la page qui les reférence référencé par la page d'accueil et qui donc hérite de leur score.


### Deuxième site : Github
```string
;extern;github.com;github.com/login;github.com/features/copilot;github.com/features/ai/github-app;github.com/mcp;github.com/features/actions;github.com/features/codespaces;github.com/features/issues;github.com/features/code-review;github.com/security/advanced-security;github.com/security/advanced-security/code-security;github.com/security/advanced-security/secret-protection;github.com/why-github;github.com/marketplace;github.com/features;github.com/enterprise
extern;0;1;1;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/login;0;1;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0
github.com/features/copilot;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features/ai/github-app;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/mcp;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features/actions;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features/codespaces;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features/issues;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features/code-review;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/security/advanced-security;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/security/advanced-security/code-security;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/security/advanced-security/secret-protection;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/why-github;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/marketplace;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/features;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
github.com/enterprise;0;1;0;1;1;1;1;1;1;1;1;1;1;1;0;1;1
```
Lien : https://github.com/

In [32]:
github = np.array([
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1],
    [0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1]
])

e = 10**(-10)
a = 0.85
Q = stochastique(github.T)
P = transition(Q, a)
r, val, i= puissance_iterée(P, e)

classement = np.argsort(-r)

print("\nMarge d'rreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
for page in classement:
    print(f"Page {page+1:<3}: {r[page]:.4f}")


Marge d'rreur : 2.080511367431169e-16

Nombre d'iteration : 4
Page 2  : 0.2949
Page 4  : 0.2645
Page 6  : 0.2645
Page 5  : 0.2645
Page 7  : 0.2645
Page 8  : 0.2645
Page 10 : 0.2645
Page 9  : 0.2645
Page 13 : 0.2645
Page 14 : 0.2645
Page 11 : 0.2645
Page 12 : 0.2645
Page 16 : 0.2645
Page 17 : 0.2645
Page 3  : 0.0358
Page 1  : 0.0339
Page 15 : 0.0339


Les pages de github sont toutes plus ou moins assez importante et surtout de meme niveau d'importance, ornis certaines pages. La page la plus visitée du site est la
page 2 qui est la page de base du site. Les pages les moins visitées sont la 15 qui est le marketplace, extern qui est la 1 et le login en 3. Le reste est très équilibré avec un nombre de liens entrant et sortant assez proche, on peut le voir avec le 0.2645 qui apparait sur toutes les pages qui n'ont pas étés cités précédemment. </br>


### Troisième site : Fruitiger Aero Archive
``` string
;extern;frutigeraeroarchive.org;frutigeraeroarchive.org/resources/foregrounds;frutigeraeroarchive.org/blog/home;frutigeraeroarchive.org/resources;frutigeraeroarchive.org/resources/userbars;frutigeraeroarchive.org/resources/buildings;frutigeraeroarchive.org/resources/trees;frutigeraeroarchive.org/resources/water;frutigeraeroarchive.org/resources/balloons;frutigeraeroarchive.org/resources/insects;frutigeraeroarchive.org/resources/people;frutigeraeroarchive.org/resources/sealife;frutigeraeroarchive.org/resources/skyboxes;frutigeraeroarchive.org/resources/clouds;frutigeraeroarchive.org/resources/bubbles
extern;0;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0
frutigeraeroarchive.org/resources/foregrounds;0;1;1;0;0;1;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/blog/home;0;1;0;0;0;0;0;0;0;0;0;0;0;0;0;0
frutigeraeroarchive.org/resources;0;0;1;0;0;1;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/userbars;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/buildings;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/trees;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/water;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/balloons;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/insects;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/people;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/sealife;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/skyboxes;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/clouds;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
frutigeraeroarchive.org/resources/bubbles;0;0;1;0;0;0;1;1;1;1;1;1;1;1;1;1
```
site internet : https://frutigeraeroarchive.org/

In [33]:
aero = np.array([
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,1,0,0,1,1,1,1,1,1,1,1,1,1,1],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,1,0,0,1,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1],
    [0,0,1,0,0,0,1,1,1,1,1,1,1,1,1,1]
    ])

e = 10**(-10)
a = 0.85
Q = stochastique(aero.T)
P = transition(Q, a)
r, val, i= puissance_iterée(P, e)

classement = np.argsort(-r)

print("\nMarge d'rreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
for page in classement:
    print(f"Page {page+1:<3}: {r[page]:.4f}")


Marge d'rreur : 1.9590477149962066e-11

Nombre d'iteration : 18
Page 3  : 0.2988
Page 8  : 0.2988
Page 7  : 0.2988
Page 13 : 0.2988
Page 12 : 0.2988
Page 11 : 0.2988
Page 10 : 0.2988
Page 9  : 0.2988
Page 15 : 0.2988
Page 14 : 0.2988
Page 16 : 0.2988
Page 2  : 0.0947
Page 6  : 0.0630
Page 4  : 0.0406
Page 5  : 0.0406
Page 1  : 0.0384


On observe que les pages de `ressources` (`buildings`, `trees`, `water`, `clouds`, etc.) obtiennent les meilleurs scores car elles forment un réseau très dense de liens internes.

**2. Dans le graphe associé, les sommets représentent des comptes et il existe une arête de `i` vers `j` lorsque le compte `i` est abonné au compte `j`.
Établir dans un premier temps le classement des comptes en fonction de leur nombre d'abonnés. Appliquer ensuite l'algorithme de PageRank version itérative et comparer le classement obtenu avec celui basé uniquement sur le nombre d'abonnés. Les écarts observés devront être expliqués en s'appuyant sur la structure du réseau et sur le rôle joué par les comptes.**



In [34]:
reseau_sociaux = np.array([
    [0,0,0,1,1,0,0,1,0,0,1,1,1,0,0,0,0,1,1,0,1,1,1,0,0],
    [0,0,0,0,0,1,1,0,1,1,0,0,0,1,1,1,1,1,0,1,1,0,0,1,1],
    [1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,1,1,1,0,1,1,0,1,0,1],
    [0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,1,1,0,0,0,1],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,1,1,0,0,1,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1],
    [0,0,1,0,1,0,0,0,0,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1],
    [0,0,0,0,0,1,1,1,1,0,0,0,0,1,1,0,1,0,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,1,1,0,0,0,1,0,0,0,0,0,0],
    [0,0,0,0,0,1,0,1,1,0,0,0,0,0,1,0,0,1,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,1],
    [0,0,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0],
    [0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,1],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [1,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0],
    [0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1],
    [0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,1,0,1,0,0,0,1],
    [0,0,0,0,0,1,1,0,0,1,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,1,1,0,0,1,0,0,1,1,0,0,0,0,0,0,0]
], dtype = 'f')

A = somme_col(reseau_sociaux)

classement = np.argsort(-A)

print("Abonné par compte")

for compte in classement:
    print(f"Page {compte+1:<3}: {A[compte]:.0f}")



Abonné par compte
Page 21 : 11
Page 20 : 11
Page 7  : 9
Page 15 : 9
Page 25 : 9
Page 18 : 8
Page 14 : 8
Page 6  : 7
Page 10 : 7
Page 9  : 7
Page 17 : 7
Page 24 : 6
Page 16 : 6
Page 19 : 3
Page 8  : 3
Page 1  : 2
Page 5  : 2
Page 12 : 2
Page 11 : 2
Page 23 : 2
Page 13 : 1
Page 4  : 1
Page 2  : 1
Page 3  : 1
Page 22 : 1


In [35]:
reseau_sociaux = np.array([
    [0,0,0,1,1,0,0,1,0,0,1,1,1,0,0,0,0,1,1,0,1,1,1,0,0],
    [0,0,0,0,0,1,1,0,1,1,0,0,0,1,1,1,1,1,0,1,1,0,0,1,1],
    [1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,1,1,1,0,1,1,0,1,0,1],
    [0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,1,1,0,0,0,1],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,1,1,0,0,1,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1],
    [0,0,1,0,1,0,0,0,0,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1],
    [0,0,0,0,0,1,1,1,1,0,0,0,0,1,1,0,1,0,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,1,1,0,0,0,1,0,0,0,0,0,0],
    [0,0,0,0,0,1,0,1,1,0,0,0,0,0,1,0,0,1,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,1],
    [0,0,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0],
    [0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,1],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [1,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0],
    [0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1],
    [0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,1,0,1,0,0,0,1],
    [0,0,0,0,0,1,1,0,0,1,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,1,1,0,0,1,0,0,1,1,0,0,0,0,0,0,0]
], dtype = 'f')


e = 10**(-10)
a = 0.85
Q = stochastique(reseau_sociaux)
P = transition(Q, a)
r, val, i= puissance_iterée(P, e)

classement = np.argsort(-r)

print("\nMarge d'rreur :", norme(r - P.dot(r)))
print("\nNombre d'iteration :", i)
for page in classement:
    print(f"Page {page+1:<3}: {r[page]:.4f}")



Marge d'rreur : 1.0631958615159334e-08

Nombre d'iteration : 52
Page 1  : 0.6500
Page 8  : 0.4180
Page 3  : 0.3892
Page 19 : 0.3130
Page 10 : 0.1692
Page 12 : 0.1594
Page 22 : 0.1313
Page 5  : 0.1256
Page 11 : 0.1203
Page 25 : 0.1030
Page 4  : 0.1021
Page 2  : 0.0963
Page 23 : 0.0612
Page 9  : 0.0538
Page 15 : 0.0530
Page 17 : 0.0439
Page 6  : 0.0420
Page 21 : 0.0410
Page 14 : 0.0363
Page 16 : 0.0347
Page 24 : 0.0332
Page 7  : 0.0326
Page 20 : 0.0283
Page 13 : 0.0246
Page 18 : 0.0229


L'écart entre les deux résultats s'explique par le fait que meme si les pages 1, 8 et 3 ont très peu d'abonnés, des pages qui elles en ont beaucoup leur sont abonnée ce qui font leur score élevé.
